Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm

```diff
@@ -0,0 +1,311 @@
+
+#if PLATFORM(IOS_FAMILY)
+#import <UIKit/UIKit.h>
+#endif
+
+#if USE(APPKIT)
+using MenuItemHandlerBlock = void (^)(id);
+
+@interface _WKWebExtensionMenuItem : NSMenuItem
+
+- (instancetype)initWithTitle:(NSString *)title handler:(MenuItemHandlerBlock)block;
+
+@property (nonatomic, copy) MenuItemHandlerBlock handler;
+
+@end
+
+@implementation _WKWebExtensionMenuItem
+
+- (instancetype)initWithTitle:(NSString *)title handler:(MenuItemHandlerBlock)handler
+{
+    ASSERT(handler);
+
+    if (!(self = [super initWithTitle:title action:@selector(_performAction:) keyEquivalent:@""]))
+        return nil;
+
+    self.target = self;
+    _handler = handler;
+
+    return self;
+}
+
+- (IBAction)_performAction:(id)sender
+{
+    _handler(sender);
+}
+
+@end
+#endif // USE(APPKIT)
+
+namespace WebKit {
+
+WebExtensionMenuItem::WebExtensionMenuItem(WebExtensionContext& extensionContext, const WebExtensionMenuItemParameters& parameters)
+    : m_extensionContext(extensionContext)
+    , m_type(parameters.type.value_or(WebExtensionMenuItemType::Normal))
+    , m_identifier(parameters.identifier)
+    , m_title(removeAmpersands(parameters.title))
+    , m_command(extensionContext.command(parameters.command))
+    , m_checked(parameters.checked.value_or(false))
+    , m_enabled(parameters.enabled.value_or(true))
+    , m_visible(parameters.visible.value_or(true))
+{
+    relaxAdoptionRequirement();
+
+    if (parameters.parentIdentifier) {
+        if (RefPtr parentMenuItem = m_extensionContext->menuItem(parameters.parentIdentifier.value()))
+            parentMenuItem->addSubmenuItem(*this);
+    }
+
+    if (parameters.contexts && !parameters.contexts.value().isEmpty())
+        m_contexts = parameters.contexts.value();
+    else if (m_parentMenuItem)
+        m_contexts = m_parentMenuItem->contexts();
+    else
+        m_contexts = WebExtensionMenuItemContextType::Page;
+
+    if (!parameters.iconDictionaryJSON.isEmpty())
+        m_icons = parseJSON(parameters.iconDictionaryJSON);
+}
+
+WebExtensionMenuItemParameters WebExtensionMenuItem::minimalParameters() const
+{
+    return {
+        identifier(),
+        parentMenuItem() ? std::optional { parentMenuItem()->identifier() } : std::nullopt,
+
+        type(),
+
+        nullString(), // title
+        nullString(), // command
+        nullString(), // iconDictionaryJSON
+
+        isChecked(),
+        isEnabled(),
+        isVisible(),
+
+        std::nullopt, // documentURLPatterns
+        std::nullopt, // targetURLPatterns
+
+        contexts(),
+    };
+}
+
+void WebExtensionMenuItem::update(const WebExtensionMenuItemParameters& parameters)
+{
+    ASSERT(extensionContext());
+
+    if (parameters.type)
+        m_type = parameters.type.value();
+
+    if (!parameters.identifier.isEmpty())
+        m_identifier = parameters.identifier;
+
+    if (parameters.parentIdentifier) {
+        RefPtr newParentMenuItem = m_extensionContext->menuItem(parameters.parentIdentifier.value());
```

Comment:
Nit: `updatedParentMenuItem`? `new` makes it sound like a new one is being created.